pkg/cri/server/sandbox_stats_list.go

```diff
@@ -0,0 +1,78 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package server
+
+import (
+	sandboxstore "github.com/containerd/containerd/pkg/cri/store/sandbox"
+	"github.com/pkg/errors"
+	"golang.org/x/net/context"
+	runtime "k8s.io/cri-api/pkg/apis/runtime/v1"
+)
+
+// ListPodSandboxStats returns stats of all ready sandboxes.
+func (c *criService) ListPodSandboxStats(
+	ctx context.Context,
+	r *runtime.ListPodSandboxStatsRequest,
+) (*runtime.ListPodSandboxStatsResponse, error) {
+	sandboxes := c.sandboxesForListPodSandboxStatsRequest(r)
+
+	podSandboxStats := new(runtime.ListPodSandboxStatsResponse)
+	for _, sandbox := range sandboxes {
+		metrics, err := metricsForSandbox(sandbox)
+
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to obtain metrics for sandbox %q", sandbox.ID)
+		}
+
+		sandboxStats, err := c.podSandboxStats(ctx, sandbox, metrics)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to decode sandbox container metrics for sandbox %q", sandbox.ID)
+		}
+		podSandboxStats.Stats = append(podSandboxStats.Stats, sandboxStats)
+	}
+
+	return podSandboxStats, nil
+}
+
+func (c *criService) sandboxesForListPodSandboxStatsRequest(r *runtime.ListPodSandboxStatsRequest) []sandboxstore.Sandbox {
+	if r.GetFilter() == nil {
```

Comment:
this doesn't feel right .. normally on these CRI list functions if there is no filter they get the whole list... pls double check

Reply:
I actually added this to be consistent with existing `ContainerStats` implementation. For container stats, in the current implementation:

1. We create a metrics request
https://github.com/containerd/containerd/blob/11ed34067ff5673e0a13d7c792acae09a4ba1120/pkg/cri/server/container_stats_list.go#L34

2. If the request is nil, we return `req, nil, nil`

https://github.com/containerd/containerd/blob/11ed34067ff5673e0a13d7c792acae09a4ba1120/pkg/cri/server/container_stats_list.go#L79-L85

3. We call `toCRIContainerStats` passing nil for `containers`

https://github.com/containerd/containerd/blob/11ed34067ff5673e0a13d7c792acae09a4ba1120/pkg/cri/server/container_stats_list.go#L42

4. Since `containers` is nil, we will ultimately skip the for loop and return an empty response

https://github.com/containerd/containerd/blob/11ed34067ff5673e0a13d7c792acae09a4ba1120/pkg/cri/server/container_stats_list.go#L57-L63

As a result, I added same logic in `SandboxStatsList`. Let me know if that makes sense. I agree it's a bit confusing, but I thought it would be good to be consistent... let me know what you think.

Reply:
nod.. thx for the discussion on slack.. 

nut of the discussion ..  let's sync with cri-o on the provided behavior for nil and empty filter list for these type requests. Don't want to spam kubelet.. also don't want the filter behavior to be different between the container runtimes.

Reply:
Thanks @mikebrow, pushed up latest revision to remove this behavior. Also filed https://github.com/containerd/containerd/issues/6198 to followup on `ListContainerStats` behavior.